src/charm.py

```diff
@@ -561,29 +583,102 @@ def _on_mongod_pebble_ready(self, event) -> None:
 
         except (PathError, ProtocolError, MissingSecretError) as e:
             logger.error("Cannot initialize workload: %r", e)
-            event.defer()
-            return
-
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("mongod", self._mongod_layer, combine=True)
-        container.add_layer("log_rotate", self._log_rotate_layer, combine=True)
+            raise FailedToUpdateFilesystem
+
+    def __configure_layers(self, container: Container) -> None:
+        """Configure the layers of the container."""
+        modified = False
+        current_layers = container.get_plan()
+        new_layers = {
+            Config.SERVICE_NAME: self._mongod_layer,
+            "log_rotate": self._log_rotate_layer,
+        }
         if self.is_role(Config.Role.CONFIG_SERVER):
-            container.add_layer("mongos", self._mongos_layer, combine=True)
+            new_layers["mongos"] = self._mongos_layer
+
+        # Add Pebble config layers missing or modified
+        for layer_name, definition in new_layers.items():
+            for service_name, service in definition.services.items():
+                if current_layers.services.get(service_name) != service:
+                    modified = True
+                    logger.debug(f"Adding layer {service_name}.")
+                    container.add_layer(layer_name, definition, combine=True)
 
+        # We'll always have a logrotate configuration at this point.
         container.exec(["chmod", "644", "/etc/logrotate.d/mongodb"])
 
-        # Restart changed services and start startup-enabled services.
-        container.replan()
+        if modified:
+            container.replan()
```

Comment:
up to you:

not 100% sure - but should replan go in the caller `_configure_container`? I would assume `configure_layers` would only create layers, but not start the service. 

up you though! just a thought I had

Reply:
It seemed weird to me to raise in `__configure_layers` when no issue happened, just nothing was changed.
So I think either:
 * replan in `__configure_layers`  
 * `__configure_layers` returns a value on which to act.

Reply:
ah - yeah you're right, sorry please keep as is!